grassroots-backend/src/phone-canvass/Scheduler/PhoneCanvassCall.ts

```diff
@@ -0,0 +1,227 @@
+/* eslint-disable grassroots/entity-use */
+import {
+  CallResult,
+  CallStatus,
+} from "grassroots-shared/dtos/PhoneCanvass/CallStatus.dto";
+import { PhoneCanvassContactEntity } from "../entities/PhoneCanvassContact.entity.js";
+import { PhoneCanvassSchedulerImpl } from "./PhoneCanvassScheduler.js";
+
+export type Call =
+  | NotStartedCall
+  | QueuedCall
+  | InitiatedCall
+  | RingingCall
+  | InProgressCall
+  | CompletedCall;
+
+interface CommonCallState {
+  id: number;
+  scheduler: PhoneCanvassSchedulerImpl;
+  contact: PhoneCanvassContactEntity;
+  // Timestamps provided by Twilio when available.
+  // Otherwise, we use Date.now().
+  transitionTimestamps: Record<CallStatus, number | undefined>;
+}
+
+type CommonCallConstructorParams = CommonCallState & {
+  currentTime: number;
+};
+
+export function resetPhoneCanvasCallIdsForTest(): void {
+  NotStartedCall.resetIdsForTest();
+}
+
+abstract class AbstractCall<STATUS extends CallStatus> {
+  readonly state: CommonCallState;
+
+  constructor(params: CommonCallState) {
+    this.state = params;
+  }
+
+  abstract get status(): STATUS;
+
+  protected advanceStatusTo<CallTypeTo extends Call>(
+    call: CallTypeTo,
+  ): CallTypeTo {
+    if (
+      !this.state.scheduler.callsByStatus[this.status].delete(this.state.id)
+    ) {
+      throw new Error(`Couldn't remove call with id ${String(this.state.id)}`);
+    }
+
+    // Sadly typescript has a rough time deducing these types.
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+    const calls = this.state.scheduler.callsByStatus[call.status] as Map<
+      number,
+      CallTypeTo
+    >;
+    calls.set(call.state.id, call);
+    this.state.scheduler.metricsTracker.onCallsByStatusUpdate(
+      this.state.scheduler.callsByStatus,
+    );
+
+    return call;
+  }
+
+  get id(): number {
+    return this.state.id;
+  }
+}
+
+export class NotStartedCall extends AbstractCall<"NOT_STARTED"> {
+  static #currentId = 0;
```

Comment:
Is the idea that every call would have a unique ID? Is there a reason this should be different from the PhoneCanvassContactEntity id?